packages/gatsby/src/datastore/lmdb/query/create-index.ts

```diff
@@ -0,0 +1,304 @@
+      : [indexFieldValue]
+
+    indexKeyElements.push(indexFieldValue)
+
+    if (indexFieldValue.length > 1) {
+      multiKeyFields.push(dottedField)
+    }
+  }
+  indexKeyElements.push([node.internal.counter])
+
+  return { keys: cartesianProduct(...indexKeyElements), multiKeyFields }
+}
+
+async function lockIndex(
+  context: IIndexingContext,
+  indexName: string
+): Promise<void> {
+  const { metadata } = context.databases
+  const indexKey = toMetadataKey(indexName)
+
+  const justLocked = await metadata.ifNoExists(indexKey, () => {
+    metadata.put(indexKey, null)
+  })
+  if (!justLocked) {
+    throw new Error(`Index is already locked`)
+  }
+}
+
+async function indexReady(
+  context: IIndexingContext,
+  indexName: string
+): Promise<IIndexMetadata> {
+  return new Promise((resolve, reject) => {
+    const { metadata } = context.databases
+
+    let retries = 0
+    let timeout = 16
+    function poll(): void {
+      const indexMetadata = metadata.get(toMetadataKey(indexName))
+      if (indexMetadata?.state === `ready`) {
+        resolve(indexMetadata)
+        return
+      }
+      if (retries++ > 1000) {
+        reject(new Error(`Index ${indexName} is locked for too long`))
+        return
+      }
+      setTimeout(poll, timeout)
+      timeout = Math.min(200, timeout * 1.5)
+    }
+    poll()
+  })
+}
+
+/**
+ * Autogenerate index name based on parameters.
+ *
+ * Example:
+ *
+ * buildIndexName(`Foo`, { foo: 1, bar: -1 }) -> `Foo/foo:1/bar:-1
+ */
+function buildIndexName(typeName: string, fields: IndexFields): string {
+  const tokens: Array<string> = [typeName]
+
+  for (const [field, sortDirection] of fields) {
+    tokens.push(`${field}:${sortDirection}`)
+  }
+
+  return tokens.join(`/`)
+}
+
+function toMetadataKey(indexName: string): string {
+  return `index:${indexName}`
+}
+
+function jsValueToLmdbKey(value: unknown): IndexFieldValue | undefined {
+  if (
+    typeof value === `number` ||
+    typeof value === `string` ||
+    typeof value === `boolean` ||
+    value === null
+  ) {
+    return value
+  }
+  if (typeof value === `undefined`) {
+    // Array keys containing `undefined` are not supported by lmdb-store
+    //  But we can't exclude those nodes from an index because
+    //  filters { eq: null, gte: null, lte: null } are expected to return such nodes
+    // Furthermore, lmdb-store puts those keys before others and we want them to be below
+    //  so need to add additional padding
+    return undefinedSymbol
+  }
+  if (Array.isArray(value)) {
+    const result: Array<IndexFieldValue> = []
+    for (const item of value) {
+      const lmdbKey = jsValueToLmdbKey(item)
+      if (typeof lmdbKey === `undefined`) {
+        return undefined // bad value
+      }
+      result.push(lmdbKey)
+    }
+    return result
+  }
+  // FIXME: not sure if we want this but there are tests for this :/
+  if (typeof value === `object`) {
+    return JSON.stringify(value)
+  }
```

Comment:
do those tests make sense? I don't think it's possible to filter or sort objects right now?

Reply:
Here are those tests (actually the whole `describe` section):

https://github.com/gatsbyjs/gatsby/blob/db0b42140b55bd8be3374b2f8b53ff834ab00da0/packages/gatsby/src/schema/__tests__/run-query.js#L2205-L2211

It tests that it correctly sorts by object value vs null, etc. Here is the example value:

https://github.com/gatsbyjs/gatsby/blob/db0b42140b55bd8be3374b2f8b53ff834ab00da0/packages/gatsby/src/schema/__tests__/run-query.js#L55

I think it was added for compatibility with `sift` and we should probably drop it / break in the strict mode. But I didn't want to mess with these tests in this PR.